src/nc/core/CxxDocument.h

```diff
@@ -0,0 +1,180 @@
+/* The file is part of Snowman decompiler. */
+/* See doc/licenses.asciidoc for the licensing information. */
+
+/* * SmartDec decompiler - SmartDec is a native code to C/C++ decompiler
+ * Copyright (C) 2015 Alexander Chernov, Katerina Troshina, Yegor Derevenets,
+ * Alexander Fokin, Sergey Levin, Leonid Tsvetkov
+ *
+ * This file is part of SmartDec decompiler.
+ *
+ * SmartDec decompiler is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * SmartDec decompiler is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with SmartDec decompiler.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <nc/config.h>
+
+#include <memory> /* std::shared_ptr */
+#include <vector>
+
+#include <boost/unordered_map.hpp>
+
+#include <nc/common/Range.h>
+#include <nc/common/Types.h>
+
+#include <QString>
+
+#include "RangeTree.h"
+
+namespace nc {
+
+namespace core {
+    class Context;
+
+    namespace arch {
+        class Instruction;
+    }
+
+    namespace ir {
+        class Statement;
+        class Term;
+    }
+
+    namespace likec {
+        class Declaration;
+        class FunctionDeclaration;
+        class FunctionDefinition;
+        class LabelDeclaration;
+        class LabelStatement;
+        class TreeNode;
+    }
+
+/**
+ * Text document containing C++ listing.
+ */
+class CxxDocument {
+    std::shared_ptr<const core::Context> context_;
+    RangeTree rangeTree_;
+    QString text_;
+    boost::unordered_map<const core::likec::TreeNode *, const RangeNode *> node2rangeNode_;
+    boost::unordered_map<const core::arch::Instruction *, std::vector<const RangeNode *>> instruction2rangeNodes_;
+    boost::unordered_map<const core::likec::Declaration *, std::vector<const core::likec::TreeNode *>> declaration2uses_;
+    boost::unordered_map<const core::likec::LabelDeclaration *, const core::likec::LabelStatement *> label2statement_;
+    boost::unordered_map<const core::likec::FunctionDeclaration *, const core::likec::FunctionDefinition *> functionDeclaration2definition_;
+
+public:
+    /**
+     * Constructor.
+     *
+     * \param context Pointer to the context. Can be nullptr.
+     */
+    CxxDocument(std::shared_ptr<const core::Context> context = nullptr);
```

Comment:
Non-GUI CxxDocument does not really need to own the context, so, the constructor should accept a raw pointer, and context_ field should go.
Then, in nocode's main.cpp the shared pointers introduced by you will go too.

The GUI CxxDocument needs shared_ptr, because different parts of the context are shown in different widgets, and different widgets are updated at different times, so, sometimes they show one context, sometimes different, so, one needs reference counting there.